lambda/matching.ml

```diff
@@ -470,405 +572,399 @@ let pretty_precompiled_res first nexts =
+              else
+                body
+          | _ -> Lstaticcatch (body, (i, []), handler) )
+    )
 
 let raw_action l =
-  match make_key l with | Some l -> l | None -> l
-
-
-let tr_raw act = match make_key act with
-| Some act -> act
-| None -> raise Exit
+  match make_key l with
+  | Some l -> l
+  | None -> l
 
 let same_actions = function
   | [] -> None
-  | [_,act] -> Some act
-  | (_,act0) :: rem ->
-      try
-        let raw_act0 = tr_raw act0 in
-        let rec s_rec = function
-          | [] -> Some act0
-          | (_,act)::rem ->
-              if raw_act0 = tr_raw act then
-                s_rec rem
-              else
-                None in
-        s_rec rem
-      with
-      | Exit -> None
-
-
-(* Test for swapping two clauses *)
-
-let up_ok_action act1 act2 =
-  try
-    let raw1 = tr_raw act1
-    and raw2 = tr_raw act2 in
-    raw1 = raw2
-  with
-  | Exit -> false
-
-let up_ok (ps,act_p) l =
+  | [ (_, act) ] -> Some act
+  | (_, act0) :: rem -> (
+      match make_key act0 with
+      | None -> None
+      | key0_opt ->
+          let same_act (_, act) = make_key act = key0_opt in
+          if List.for_all same_act rem then
+            Some act0
+          else
+            None
+    )
+
+let safe_before (ps, act_p) l =
+  (* Test for swapping two clauses *)
+  let same_actions act1 act2 =
+    match (make_key act1, make_key act2) with
+    | Some key1, Some key2 -> key1 = key2
+    | None, _
+    | _, None ->
+        false
+  in
   List.for_all
-    (fun (qs,act_q) ->
-      up_ok_action act_p act_q || not (may_compats ps qs))
+    (fun (qs, act_q) -> same_actions act_p act_q || not (may_compats ps qs))
     l
 
 (*
-   The simplify function normalizes the first column of the match
+   The half-simplify functions transforms the first column of the match
      - records are expanded so that they possess all fields
      - aliases are removed and replaced by bindings in actions.
-   However or-patterns are simplified differently,
-     - aliases are not removed
-     - or-patterns (_|p) are changed into _
-*)
 
-exception Var of pattern
-
-let simplify_or p =
-  let rec simpl_rec p = match p with
-    | {pat_desc = Tpat_any|Tpat_var _} -> raise (Var p)
-    | {pat_desc = Tpat_alias (q,id,s)} ->
-        begin try
-          {p with pat_desc = Tpat_alias (simpl_rec q,id,s)}
-        with
-        | Var q -> raise (Var {p with pat_desc = Tpat_alias (q,id,s)})
-        end
-    | {pat_desc = Tpat_or (p1,p2,o)} ->
-        let q1 = simpl_rec p1 in
-        begin try
-          let q2 = simpl_rec p2 in
-          {p with pat_desc = Tpat_or (q1, q2, o)}
-        with
-        | Var q2 -> raise (Var {p with pat_desc = Tpat_or (q1, q2, o)})
-        end
-    | {pat_desc = Tpat_record (lbls,closed)} ->
+   However or-patterns are only half-simplified,
+     - aliases under or-patterns are kept
+     - or-patterns whose left-hand-side is simplifiable
+       are removed: (_|p) is changed into _
```

Comment:
Here, the code feels less specific than the comment since any or-pattern `p|q` where `p ≤ q` is simplified to `p`. Maybe we could add a `for instance` in the comment, or write down the effective rule. Or maybe, 

 > - or-patterns whose right-hand-side is subsumed by their left-hand side are simplified to their left-hand-side:
       For instance  [( _::_ | 1 :: _  )] is changed into [ _ :: _ ]